src/librustdoc/clean/types.rs

```diff
@@ -1394,33 +1408,50 @@ impl Type {
     }
 }
 
-impl GetDefId for Type {
-    fn def_id(&self) -> Option<DefId> {
+impl Type {
+    fn inner_def_id(&self, cache: Option<&Cache>) -> Option<DefId> {
+        fn inner<T: GetDefId>(t: &T, cache: Option<&Cache>) -> Option<DefId> {
+            match cache {
+                Some(c) => t.def_id_full(c),
+                None => t.def_id(),
+            }
+        }
+
         match *self {
             ResolvedPath { did, .. } => Some(did),
-            Primitive(p) => cache().primitive_locations.get(&p).cloned(),
+            Primitive(p) => cache.and_then(|c| c.primitive_locations.get(&p).cloned()),
             BorrowedRef { type_: box Generic(..), .. } => {
-                Primitive(PrimitiveType::Reference).def_id()
+                inner(&Primitive(PrimitiveType::Reference), cache)
             }
-            BorrowedRef { ref type_, .. } => type_.def_id(),
+            BorrowedRef { ref type_, .. } => inner(&**type_, cache),
             Tuple(ref tys) => {
                 if tys.is_empty() {
-                    Primitive(PrimitiveType::Unit).def_id()
+                    inner(&Primitive(PrimitiveType::Unit), cache)
                 } else {
-                    Primitive(PrimitiveType::Tuple).def_id()
+                    inner(&Primitive(PrimitiveType::Tuple), cache)
                 }
             }
-            BareFunction(..) => Primitive(PrimitiveType::Fn).def_id(),
-            Never => Primitive(PrimitiveType::Never).def_id(),
-            Slice(..) => Primitive(PrimitiveType::Slice).def_id(),
-            Array(..) => Primitive(PrimitiveType::Array).def_id(),
-            RawPointer(..) => Primitive(PrimitiveType::RawPointer).def_id(),
-            QPath { ref self_type, .. } => self_type.def_id(),
+            BareFunction(..) => inner(&Primitive(PrimitiveType::Fn), cache),
+            Never => inner(&Primitive(PrimitiveType::Never), cache),
+            Slice(..) => inner(&Primitive(PrimitiveType::Slice), cache),
+            Array(..) => inner(&Primitive(PrimitiveType::Array), cache),
+            RawPointer(..) => inner(&Primitive(PrimitiveType::RawPointer), cache),
+            QPath { ref self_type, .. } => inner(&**self_type, cache),
             _ => None,
```

Comment:
It doesn't need to be fixed here, but it would be nice to remove this wildcard.

Reply:
Adding a note then so we can do it afterwards.